renku/core/migration/m_0010__metadata_fixes.py

```diff
@@ -0,0 +1,310 @@
+                # NOTE: Fix invalidation times set before creation date on plans
+                plan.unfreeze()
+                plan.invalidated_at = plan.date_created
+                plan.freeze()
+            if plan.invalidated_at.tzinfo is None:
+                # NOTE: There was a bug that caused invalidated_at to be set without timezone (as UTC time)
+                # so we patch in the timezone here
+                plan.unfreeze()
+                plan.invalidated_at = plan.invalidated_at.replace(microsecond=0).astimezone(timezone.utc)
+                plan.freeze()
+        if plan.id in activity_map and plan.date_created > activity_map[plan.id].started_at_time:
+            plan.unfreeze()
+            plan.date_created = activity_map[plan.id].started_at_time
+            plan.freeze()
+
+    # NOTE: switch creation date for modification date
+    for tail in plan_gateway.get_newest_plans_by_names(include_deleted=True).values():
+        stack: List[AbstractPlan] = []
+        stack.append(tail)
+        creation_date = tail.date_created
+        plan = tail
+
+        while plan.is_derivation():
+            plan = cast(AbstractPlan, plan_gateway.get_by_id(plan.derived_from))
+            creation_date = plan.date_created
+            stack.append(plan)
+
+        while stack:
+            plan = stack.pop()
+            plan.unfreeze()
+            plan.date_modified = plan.date_created
+            plan.date_created = creation_date
+            plan.freeze
+
+    database.commit()
+
+
+@inject.autoparams("dataset_gateway")
+def fix_dataset_date_modified(dataset_gateway: IDatasetGateway):
+    """Change date_created and date_modified to have correct semantics."""
+    tails = dataset_gateway.get_provenance_tails()
+
+    for dataset_tail in tails:
+        dataset = dataset_tail
+        found_datasets: List[Dataset] = []
+        previous_modification_date = local_now()
+
+        while dataset.derived_from is not None:
+            modification_date = dataset.date_removed or dataset.date_created
+
+            if modification_date is not None:
+                assert modification_date < previous_modification_date
+                dataset.unfreeze()
+                dataset.date_modified = modification_date
+                dataset.freeze()
+                previous_modification_date = modification_date
+            found_datasets.append(dataset)
+            dataset = dataset_gateway.get_by_id(dataset.derived_from.value)
+
+        # NOTE: first dataset in chain
+        modification_date = dataset.date_created or dataset.date_published
+        if modification_date is not None:
+            dataset.unfreeze()
+            dataset.date_modified = modification_date
+            dataset.freeze()
+
+        for child in found_datasets:
+            child.unfreeze()
+            child.date_created = dataset.date_created
+            child.date_published = dataset.date_published
```

Comment:
Derivative datasets shouldn't have a `date_published` so we should remove this and modify the upper statement to `child.date_created = dataset.date_created or dataset.date_published` or just `child.date_created = modification_date`